### ISSUE TITLE
Maximum number of failures or failure backoff policy for Jobs

### ISSUE THREAD

**Issue by maximz** [kubernetes/kubernetes#30243]
A max number of failures or failure backoff policy for Jobs would be useful.

Imagine you have an ETL job in production that's failing due to some pathological input.
By the time you spot it, it's already been rescheduled thousands of times. As an example, I had 20 broken jobs that kept getting rescheduled; killing them took forever -- and crashed the Kubernetes dashboard in the process.

Today, "restartPolicy" is not respected by Jobs because the goal is to achieve successful completion. (Strangely, "restartPolicy: Never" is still valid YAML.) This means failed jobs keep getting rescheduled. When you go to delete them, you have to delete all the pods they've been scheduled on. Deletes are rate limited, and in v1.3+, the verbose "you're being throttled" messages are hidden from you when you run the `kubectl` command to delete a job. So it just looks like it's taking forever! This is not a pleasant UX if you have a runaway job in prod or if you're testing out a new job and it takes minutes to clean up after a broken test.

What are your thoughts on specifying a maximum number of failures/retries or adding a failure backoff restart policy?

(cc @thockin , who [suggested I file this feature request](https://twitter.com/zazius/status/758816688171126784))

---

Ninja edit: [per this SO thread](http://stackoverflow.com/questions/35141607/is-there-a-max-retries-for-kubernetes-jobs), the throttling issue is avoidable by using the OnFailure restart policy to keep rescheduling to the same pods, rather than to new pods -- i.e. this prevents the explosion of the number of pods. And deadlines can help weed out failures after a certain amount of time.

However, suppose my ETL job takes an hour to run properly but may fail within seconds if the input data is bad. I'd rather specify a maximum number of retries than a high deadline.

**Comment by cioc** [comment]
@thockin , @maximz  re your twitter convo, the docs (http://kubernetes.io/docs/user-guide/jobs/) imply that restartPolicy = Never is supported for jobs, which I found misleading.  

1.  The first example shows that restartPolicy = Never 
2.  The Pod Template section reads, "Only a RestartPolicy equal to Never or OnFailure are allowed."
3. The Handling Pod and Container failures section reads, "Therefore, your program needs to handle the the case when it is restarted locally, or else specify .spec.template.containers[].restartPolicy = "Never""

If restartPolicy = Never shouldn't be allowed and the goal here is that jobs run until completion, then the docs need to change along with the feature request of max failures / retries.

**Comment by soltysh** [comment]
@erictune and I authored Jobs, I'll be out for the next two weeks but here are my quick thoughts. Currently `restartPolicy` does not apply to Jobs almost at all, it's passed over to kubelet when creating pods. To address your concern we'd have to come up with a policy for a job which would allow setting logic around how job should react to failures. Overall, I'd like to see such a feature. Not sure how Eric sees that?

**Comment by sstarcher** [comment]
This is a much needed feature for us as we have several things we would like to run via the Jobs structure, but we cannot as they would get restarted on failure which is not desirable in a few situations for us.

**Comment by erictune** [comment]
I support adding a backoff policy to Jobs. I think exponential backoff with a max of like 5min would be fine.   I don't think it would be a breaking change to introduce a backoff policy by default.  I suspect some users might want fast retry, but we can wait on adding a field until they do.

I also support having a "max failed pods to keep around", which would cause the job controller to garbage collect some failed pods before creating new ones.  At a minimum, keeping the first and the last failed pod would be useful for debugging.  But keeping like 1000 failed pods is usually not useful.  Especially if parallelism is 1.  I'm not sure if we can change this to be a default, but we can definitely make it a knob.

I'd want to discuss it a bit more, but I'd also be open to a field whose meaning is "the job can move to a failure state after this many pod failures".  We already have a way to fail after a certain amount of time.

I am not very enthusiatic about a "max pod failures before job fails" feature.  In particular, I don't  know that we can easily guarantee that a job only ever tries one time.

**Comment by erictune** [comment]
In this comment: https://github.com/kubernetes/kubernetes/issues/24533#issuecomment-217058534 Brian Grant says there should be backoff with no knob.

**Comment by erictune** [comment]
A default pods object quota per namespace would also help protect the apiserver from ending up with too many jobs.  Users could always raise their quota if needed.

**Comment by hanikesn** [comment]
As I understand the pod object limit only applies to non-terminal pods:

>The total number of pods in a non-terminal state that can exist in the namespace. A pod is in a terminal state if status.phase in (Failed, Succeeded) is true.
http://kubernetes.io/docs/admin/resourcequota/

**Comment by soltysh** [comment]
> And currently there's no way to limit the total number of pods. But I guess this would be another issue.

The completed pods should be cleared by the gc at some point in time.

**Comment by diegodelemos** [comment]
What is the current state of this issue? We are really interested in having such behaviour because we are going to use Kubernetes to execute a set of steps of a workflow and if a certain step fails, what we would expect is: the job stops and a option to get a log specifying the failure reason is available.

Besides, addressed to @erictune on behalf of his [comment](https://github.com/kubernetes/kubernetes/issues/30243#issuecomment-244216231) about jobs and the apiserver. What would be considered too many jobs for the apiserver? We are basing the execution of the steps of several workflows on Kubernetes.

**Comment by soltysh** [comment]
@diegodelemos you're open to submit a patch with an appropriate fix at any time :)

**Comment by Yancey1989** [comment]
Hi @diegodelemos @soltysh,  how about this issue? I also have many demands about this.

**Comment by soltysh** [comment]
@Yancey1989 you're welcome to submit a patch as well 😃

**Comment by bgrant0607** [comment]
@soltysh Are you working on this?

cc @janetkuo

**Comment by soltysh** [comment]
@bgrant0607 I'd wish, but I'm out of time :( Just recently somebody opened https://github.com/kubernetes/kubernetes/pull/41451 but that is completely not what we want. I asked the original submitter to discuss proposed solution here, but haven't heard from him since.

**Comment by krabhishek8260** [comment]
@bgrant0607 @soltysh Is anyone working on this?

**Comment by nickschuch** [comment]
@soltysh, can you please elaborate on "completely not what we want"? From what I saw in that ticket is sounded like we just need extra toggles for additional cases. I have a pretty clear outline of what Im trying to achieve in that PR (if you want me to duplicate it in here I can). Im happy to work on this, Im just not to sure how to proceed.

**Comment by soltysh** [comment]
@nickschuch first of all thanks for you PR and your interest! As for what I say, your solution although is perfectly fine ad-hoc, we're interested in having much more structural approach to this problem. Having just int to count failures is definitely not sufficient. Have a look at [this comment of mine](https://github.com/kubernetes/kubernetes/pull/41451#discussion_r101523399), lemme copy that over here:

> 1. count failures from beginning, iow. I don't care when the failures started, but as soon as they hit X we can fail a job
> 2. count failures from certain point (probably configurable, eg. X successful), and Y failures after that fails the job
> 3. be able to specify whether to clear the counter if successful run happens in the mean time

 We need to have an agreement on that approach first and only then dive into implementation. I'd prefer to have it addressed once and for all (the API part, not the implementation), and not one step at a time, because this might cause problems later on. Especially that we're thinking of moving CronJobs to beta in 1.7 which will make it harder to modify API afterwards. Hope that's reasonable?

**Comment by soltysh** [comment]
@nickschuch will you be able to commit yourself and have it done in time for 1.7?

**Comment by soltysh** [comment]
> Especially that we're thinking of moving CronJobs to beta in 1.7 which will make it harder to modify API afterwards.

That's obviously wrong, since we're talking about Jobs here, and not CronJobs (like in #43512). So changing APIs here is even more tricky and that's why I'm insisting on having the discussion first.

**Comment by nickschuch** [comment]
> @nickschuch will you be able to commit yourself and have it done in time for 1.7?

Yep, absolutely. I've been looking for where to best contribute my time on K8s. I want to see this through.

> I'd prefer to have it addressed once and for all (the API part, not the implementation), and not one step at a time, because this might cause problems later on.

I agree, it will make it easier for me to understand this issue.

How do we go about getting agreement? (sorry, still new to development in the K8s community)

**Comment by soltysh** [comment]
Sorry, was out last week.

> How do we go about getting agreement? (sorry, still new to development in the K8s community)

Open a PR updating [the current job proposal](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/job.md) with your proposed changes. You can also start shaping the proposal here.

**Comment by lukasheinrich** [comment]
hi @soltysh @nickschuch a bunch of people on our side (CERN) are also interested in this ( @diegodelemos @rochaporto) and we'd be interested to help.

**Comment by nickschuch** [comment]
No worries, I've blocked out a couple of hours for tomorrow, will have a first crack at the proposal and go from there.

@lukasheinrich that would be awesome!

**Comment by soltysh** [comment]
@lukasheinrich awesome, the more people will give feedback now when we're shaping the design the better!

**Comment by soltysh** [comment]
I've created this proposal to address the issue: https://github.com/kubernetes/community/issues/583

**Comment by sdminonne** [comment]
@soltysh any chances to have this implemented?  This starts to bite us too many times :)
so... +1

**Comment by soltysh** [comment]
I'll do my best, but I can't promise anything, b/c I'm traveling this and next week.

**Comment by nickschuch** [comment]
We are still waiting for this proposal to be agreed upon right? Im happy to write the code (written some already), not very familiar with the procedures sorry

**Comment by soltysh** [comment]
@sdminonne it looks we're not gonna make it :( sorry

**Comment by PeiKevin** [comment]
How about the status about this issue? I'm really want to use this feature, If  I want to run a job, I just want to know success or failure with no more try. if always retrying, it will always be failed.

**Comment by soltysh** [comment]
API PR: https://github.com/kubernetes/kubernetes/pull/48075
Controller PR: https://github.com/kubernetes/kubernetes/pull/51153
